Favorites with current weather

teresa-m-knowles opened the pull request:
Closes #8, #9 and #10 

A user can get a list of their favorite locations with their current weather. 

<img width="1002" alt="Screen Shot 2019-05-04 at 5 03 09 PM" src="https://user-images.githubusercontent.com/13354855/57199808-c11cf000-6f40-11e9-8687-f9ed38ff530a.png">

Also here's the parent class Forecast from which FavoriteFormat inherits from: 
<img width="847" alt="Screen Shot 2019-05-05 at 2 29 29 PM" src="https://user-images.githubusercontent.com/13354855/57199929-3a691280-6f42-11e9-9742-eb98086ac444.png">

teresa-m-knowles committed:
Skeleton of the get request for locations

teresa-m-knowles committed:
Adds headers to response

teresa-m-knowles committed:
Creates a FavoriteFormat class that inherits from Forecast

teresa-m-knowles committed:
Adds a get request to see a user's favorite locations

teresa-m-knowles committed:
Uses i instead of counter and fixes the catch error

aprildagonese commented on routes/api/v1/favorites.js:
```diff
@@ -54,5 +58,54 @@ router.post("/", function(req, res) {
+            for (i = 0; i < locations.length; i++) {
+              let currentLocation = locations[i];
+              let forecastUrl = `https://api.darksky.net/forecast/${process.env.DARK_SKY_API_KEY}/${currentLocation.lat},${currentLocation.lng}?exclude=[minutely,alerts,flags]`;
+              fetch(forecastUrl)
```
If you have time, you could think about breaking any fetch calls out into services, like we did with the original SweaterWeather, to clean this up a bit.

aprildagonese commented on routes/api/v1/favorites.js:
```diff
@@ -54,5 +58,54 @@ router.post("/", function(req, res) {
+        user.getLocations()
+          .then(locations => {
+            var userLocations = locations;
+            for (i = 0; i < locations.length; i++) {
```
Instead of a counter, you could try using map to iterate through all your locations and produce a new array. Might make this slightly more readable!

teresa-m-knowles replied:
Great idea, thanks! I did locations.map which I think it's easier to understand

aprildagonese commented on routes/api/v1/favorites.js:
```diff
@@ -54,5 +58,54 @@ router.post("/", function(req, res) {
+                    res.status(200).send(JSON.stringify(favorites));
+                  }
+                })
+                .catch(error => console.log(error));
```
These different errors get a bit hard to read for somebody new to your code. One suggestion is to use more explicit error names. Or if you don't want to change them, you could even just add a few comments to differentiate them, since JS seems more forgiving of comments than Ruby.

teresa-m-knowles replied:
Changed the error names to be more user friendly

aprildagonese commented on routes/api/v1/favorites.js:
```diff
@@ -54,5 +58,54 @@ router.post("/", function(req, res) {
+    include: [Location]
+  })
+    .then( user => {
+      if(!user){
```
If you wanted to cut back on a couple lines here, the ternary operator is pretty for simple if/else conditions. It would be like:
`!user ? res.sendStatus(401) :`
and then your "else" clause would start on the next line.

teresa-m-knowles replied:
Thank you! Great use of that operator, just changed it

aprildagonese commented on helpers/favorite_format.js:
```diff
@@ -0,0 +1,15 @@
+var Forecast = require('./forecast')
+
+
+module.exports = class FavoriteFormat extends Forecast{
```
As Cory said the other day, is inheritance a good idea or a bad idea here? I don't have an opinion, just wondered what your thought process was :)

aprildagonese reviewed (commented):
Made some suggestions for a couple areas that could help with readability, but overall what you're doing here makes sense. Looks good Teresa!

teresa-m-knowles committed:
Changes if/else in get favorites to use ternary operator instead

teresa-m-knowles committed:
Changes from for loop to map method

teresa-m-knowles commented:
Thanks @aprildagonese ! Great suggestions. Per your comments, I: 

- changed from if/else to a ternary operator
- added better error messages 
- changed from a for loop to iterating through the user locations with locations.map

I am going to finish all the endpoints before refactoring things into a service object, that's a great idea though.

Thanks for being so thorough
